torch/csrc/autograd/VariableTypeManual.cpp

```diff
@@ -152,6 +152,37 @@ Tensor& requires_grad_(Tensor& self, bool _requires_grad) {
   return self.set_requires_grad(_requires_grad);
 }
 
+void retain_grad(const Tensor & self) {
+  TORCH_CHECK(self.requires_grad(), "can't retain_grad on Tensor that has requires_grad=False");
+  if (self.is_leaf()) {  // no-op for leaves
+    return;
+  }
+  if (impl::get_autograd_meta(self)->retains_grad_) {
+    return;
+  }
+  c10::weak_intrusive_ptr<TensorImpl> weak_self(self.getIntrusivePtr());
+
+  std::function<void(Tensor)> retain_grad_hook([weak_self](const Tensor& grad) {
```

Comment:
does it actually need to be a weak pointer? Can it be a raw pointer to TensorImpl?

What is the situation when weak ptr might expire here? Afaiu, the hook is being referenced by the owning Tensor (that's why we don't want to create a circular reference).

However, it seems that hooks in AutogradMeta a shared_ptr: https://github.com/pytorch/pytorch/blame/48569cc33047fb2badf3634ebb439c0c54d45d7a/torch/csrc/autograd/variable.h#L197 - does it mean the they can be shared between several tensors? If we could make that unique_ptr, then the Tensor would have unique ownership of the hook and in this code here we could do just raw TensorImpl*.

cc @albanD @swolchok

Reply:
> What is the situation when weak ptr might expire here?

I guess if you delete the Tensor before calling the backward.
The idea here is that in that case, since the user cannot access that Tensor anymore, there is no need to populate its `.grad` field.

The hook here is actually kept alive by the `Node` in the graph (not the Tensor). And so the hook can outlive the Tensor.

> the hook is being referenced by the owning Tensor

This is not strictly true. The hook is referenced via the `Node` in the graph.
So yes not having a weak ptr would create a ref cycle but the hook can still outlive the Tensor.

Does that answer your question?